client/src/websockets.rs

```diff
@@ -0,0 +1,632 @@
+        }
+
+        let mut header_vec: Vec<String> = Vec::new();
+
+        if let Some(headers) = options.headers {
+            headers
+                .iter()
+                .map(|(key, value)| header_vec.push(format!("{}:{}", key, value)));
+        }
+
+        //Make the protobuf message and convert it to bytes
+        let ws_msg = WebSocketMessage {
+            r#type: Some(1),
+            request: Some(WebSocketRequestMessage {
+                id: Some(id),
+                verb: Some(options.verb),
+                path: Some(options.path),
+                headers: header_vec,
+                body: options.body,
+            }),
+            response: None,
+        };
+        let msg = Message::Binary(ws_msg.encode_to_vec());
+
+        //Send the message over ws
+        self.socket.lock().await.stream.send(msg).await?;
+
+        let mut self_clone = self.clone();
+
+        let promise = spawn(async move {
+            if self_clone.has_keepalive {
+                set_timeout(KEEPALIVE_TIMEOUT_MS.into(), || async move {
+                    self_clone
+                        .outgoing_map
+                        .lock()
+                        .await
+                        .remove(&id)
+                        .unwrap()
+                        .abort();
+                    self_clone
+                        .socket
+                        .lock()
+                        .await
+                        .stream
+                        .send(Message::Close(Some(frame::CloseFrame {
+                            code: 3008.into(),
+                            reason: "Timed out".into(),
+                        })))
+                        .await;
+                    //TODO: Kill the listener process
+                })
+                    .await;
+            }
+        });
+
+        //Add future to hashmap
+        self.outgoing_map
+            .lock()
+            .await
+            .insert(id, promise.abort_handle());
+
+        promise.await;
+
+        if let Some(incoming) = self.incoming_response.lock().await.get(&id) {
+            let status = match incoming.status {
+                Some(status) => status,
+                _ => panic!(),
+            };
+
+            if (200..300).contains(&status) {
+                Ok(incoming.clone())
+            } else {
+                anyhow::bail!("Got a bad response")
+            }
+        } else {
+            anyhow::bail!("Incoming response missing")
+        }
+    }
+
+    //TODO: Send TLS close_notify when closing connection
+    pub async fn close(&mut self, code: u16, reason: String) -> Result<()> {
+        Ok(self
+            .socket
+            .lock()
+            .await
+            .stream
+            .close(Some(frame::CloseFrame {
+                code: code.into(),
+                reason: reason.into(),
+            }))
+            .await?)
+    }
+}
+
+pub struct SendRequestOptions {
```

Comment:
why not use the websocketmessage that is already defined?

Reply:
It's a copy of the Signal-Desktop behavior where send_request has an options parameter